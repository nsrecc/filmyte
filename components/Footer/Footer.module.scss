@use "styles/variables" as *;

.footer {
  width: 100%;
  height: 100px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: $ui-color-background-secondary;
  
  a {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  // all lists
  ul {
    display: inline-flex;
    list-style: none;
    padding: 0;

    // all list items
    li {
      padding: 0.5em 2rem;
      border-right: 1px solid $gray-50;
      display: flex;
      justify-content: center;
      align-items: center;

      &:last-of-type {
        border-right: 0;
      }
    }

    // "powered by" sub list items
    .poweredBy {
      > li {
        padding: 0.5rem 1rem;

        &:last-of-type {
          padding-right: 0;
        }
      }
    }
  }
}

.logo {
  height: 1em;
}
