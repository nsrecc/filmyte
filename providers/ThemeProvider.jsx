import React, {
  useState, useEffect, useContext, createContext,
} from 'react';
import PropTypes from 'prop-types';
import includes from 'lodash/includes';
import styles from './ThemeProvider.module.scss';

// List of accepted themes
const THEME_LIST = ['light', 'dark'];

// Create ThemeContext with default object to be accessed in any nested component as needed.
const ThemeContext = createContext({
  theme: 'light',
  themeIsLight: true,
  toggleTheme: () => {},
});

// Export useThemeContext to retrieve theme state object via useContext hook at any nested
// components in order to update ThemeContext theme state there.
export const useThemeContext = () => useContext(ThemeContext);

/**
 * Custom hook useLocalTheme
 * - Create theme state with defaultTheme
 * - Updates theme in local storage and state
 * - Gets theme when ThemeProvider mounts
 * @param {string} defaultTheme - default theme, should be 'light'
 */
const useLocalTheme = (defaultTheme) => {
  // create theme state
  const [theme, setTheme] = useState(defaultTheme);

  // take new theme and set in it local storage, update theme state
  const setThemeLocally = (newTheme) => {
    localStorage.setItem('theme', newTheme);
    setTheme(newTheme);
  };

  // when ThemeProvider is mounted, get theme from local storage and set it to theme state
  useEffect(() => {
    const retrievedLocalTheme = localStorage.getItem('theme');

    // ensure only accepted themes can be set to theme state
    if (includes(THEME_LIST, retrievedLocalTheme)) setTheme(retrievedLocalTheme);
  }, []);

  return [theme, setThemeLocally];
};

/**
 * ThemeProvider wraps the application in _app.jsx with passed values:
 * - theme state
 * - theme check
 * - toggle theme function
 */
export const ThemeProvider = ({ children }) => {
  const [theme, setThemeLocally] = useLocalTheme('light');

  const themeIsLight = theme === 'light';

  // sets new theme using setThemeLocally from custom hook useLocalTheme
  const toggleTheme = () => {
    return themeIsLight ? setThemeLocally('dark') : setThemeLocally('light');
  };

  // div element here should contain 'app' style class as it's a direct parent to <main>
  return (
    <ThemeContext.Provider value={{ theme, themeIsLight, toggleTheme }}>
      <div className={`app ${styles.theme} ${themeIsLight ? styles.light : styles.dark}`}>
        {children}
      </div>
    </ThemeContext.Provider>
  );
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
