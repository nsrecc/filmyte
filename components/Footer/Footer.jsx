import React from 'react';
import { useThemeContext } from 'providers/ThemeProvider/ThemeProvider';
import styles from './Footer.module.scss';

const Footer = () => {
  const { themeIsLight } = useThemeContext();

  const vercelSvgSrc = themeIsLight ? '/icons/vercel.svg' : '/icons/vercel-white.svg';

  return (
    <footer className={styles.footer}>
      <div className={styles.footerInner}>
        <p id="copyright">&copy; 2021 filmyte.com</p>
        <div className={styles.poweredBy}>
          <p className={styles.poweredByText}>Powered by:</p>
          <ul className={styles.poweredByList}>
            <li>
              <a href="https://www.themoviedb.org/" target="_blank" rel="noopener noreferrer">
                The Movie Database (TMDb)
              </a>
            </li>
            <li>
              <a href="https://vercel.com/" target="_blank" rel="noopener noreferrer">
                <img
                  className={styles.vercelLogo}
                  id="vercel-logo"
                  alt="Vercel Logo"
                  src={vercelSvgSrc}
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </footer>
  );
};

export default Footer;
