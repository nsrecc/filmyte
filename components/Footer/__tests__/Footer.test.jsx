import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { useThemeContext } from 'providers/ThemeProvider/ThemeProvider';
import Footer from '../Footer';

jest.mock('providers/ThemeProvider/ThemeProvider', () => ({
  useThemeContext: jest.fn().mockName('useThemeContext'),
}));

// TO DO: test for existing TMDb logo
describe('Footer', () => {
  beforeEach(() => {
    useThemeContext.mockReturnValue({ themeIsLight: true });
  });

  // after each test, clear the calls and instances of all mocks
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<Footer />);
    expect(wrapper).toMatchSnapshot();
  });

  it('contains copyright notice', () => {
    const wrapper = shallow(<Footer />);
    expect(wrapper.find('#copyright').exists()).toBe(true);
  });

  describe('Footer with light theme', () => {
    it('contains vercel logo with color black', () => {
      const wrapper = shallow(<Footer />);
      expect(wrapper.find('#vercel-logo').exists()).toBe(true);
      expect(wrapper.find('#vercel-logo').prop('src')).toBe('/icons/vercel.svg');
    });
  });

  describe('Footer with dark theme', () => {
    beforeEach(() => {
      useThemeContext.mockReturnValue({ themeIsLight: false });
    });

    it('contains vercel logo with color white', () => {
      const wrapper = shallow(<Footer />);
      expect(wrapper.find('#vercel-logo').exists()).toBe(true);
      expect(wrapper.find('#vercel-logo').prop('src')).toBe('/icons/vercel-white.svg');
    });
  });
});
