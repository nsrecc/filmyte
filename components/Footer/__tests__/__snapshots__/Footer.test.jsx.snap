// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer matches snapshot 1`] = `
<footer
  className="footer"
>
  <div
    className="footerInner"
  >
    <p
      id="copyright"
    >
      Â© 2021 filmyte.com
    </p>
    <div
      className="poweredBy"
    >
      <p
        className="poweredByText"
      >
        Powered by:
      </p>
      <ul
        className="poweredByList"
      >
        <li>
          <a
            href="https://www.themoviedb.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            The Movie Database (TMDb)
          </a>
        </li>
        <li>
          <a
            href="https://vercel.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="Vercel Logo"
              className="vercelLogo"
              id="vercel-logo"
              src="/icons/vercel.svg"
            />
          </a>
        </li>
      </ul>
    </div>
  </div>
</footer>
`;
